### PR TITLE
EPMDJ-5963 and EPMDJ-6065 Fix start session for offline agent and fix multinstances agent is busy when one of nodes is up

### PR DESCRIPTION
**Pull request by Mazel-Tovr**

### CODE REVIEW

File: core/src/main/kotlin/endpoints/AgentManager.kt [hunk 1]
```diff
@@ -217,13 +216,14 @@ class AgentManager(override val kodein: Kodein) : KodeinAware {
...
                 description = dto.description,
-                status = AgentStatus.BUSY,
+                isRegistered = true,
                 plugins = pluginsToAdd.mapTo(mutableSetOf()) { it.pluginBean.id }
             )
         }
         adminData(agentId).updateSettings(dto.systemSettings)
         pluginsToAdd.forEach { plugin -> ensurePluginInstance(entry, plugin) }
         info.sync()
+        info.persistToDatabase()
```
Comment (Mazel-Tovr) [inline core/src/main/kotlin/endpoints/AgentManager.kt]: i should persist after copy

File: core/src/main/kotlin/endpoints/AgentManager.kt [hunk 2]
```diff
@@ -550,10 +534,64 @@ class AgentManager(override val kodein: Kodein) : KodeinAware {
...
+        updateInstanceStatus(agentId, instanceId, AgentStatus.BUSY).also { instance ->
+            if (instance?.all { it.value.status == AgentStatus.BUSY } == true) {
+                notifyAgents(agentId)
+            }
+        }
+        try {
+            block(instanceInfo.agentWsSession)
+        } finally {
+            updateInstanceStatus(agentId, instanceId, AgentStatus.ONLINE).also { instance ->
+                if (instance?.all { it.value.status != AgentStatus.BUSY } == true) {
```
Comment (LikhanovMaxim) [inline core/src/main/kotlin/endpoints/AgentManager.kt]: smth strange condition

File: tests/src/test/kotlin/e2e/AgentSingleInstanceTest.kt [hunk 3]
```diff
@@ -41,7 +41,11 @@ class AgentSingleInstanceTest : E2ETest() {
                     status shouldBe AgentStatus.ONLINE
                     instanceIds.size shouldBe 1
                 }
-            }.reconnect(aw) { ui, _ ->
+            }.reconnect(aw) { ui, agent ->
+                ui.getAgent()?.status shouldBe AgentStatus.ONLINE
+                ui.getAgent()?.status shouldBe AgentStatus.BUSY
+                agent.`get-set-packages-prefixes`()
+                agent.`get-load-classes-datas`()
```
Comment (ViktoriiaLebedeva) [inline tests/src/test/kotlin/e2e/AgentSingleInstanceTest.kt]: Why did you do this only in this test?

**Review by ViktoriiaLebedeva — CHANGES_REQUESTED** [review]

File: core/src/main/kotlin/endpoints/AgentManager.kt [hunk 4]
```diff
@@ -549,10 +532,62 @@ class AgentManager(override val kodein: Kodein) : KodeinAware {
...
+        agentId: String,
+        instanceId: String,
+        instanceInfo: InstanceInfo,
+        block: suspend AgentWsSession.() -> Unit
+    ) {
+        updateInstanceStatus(agentId, instanceId, AgentStatus.BUSY).also { instance ->
+            if (instance?.all { it.value.status == AgentStatus.BUSY } == true) {
+                notifyAgents(agentId)
+            }
+        }
```
Comment (Mazel-Tovr) [inline core/src/main/kotlin/endpoints/AgentManager.kt]: add logging

File: core/src/main/kotlin/endpoints/AgentManager.kt [hunk 5]
```diff
@@ -568,3 +603,8 @@ private data class InstancesKey(
     val agentId: String,
     val buildVersion: String,
 )
+
+data class InstanceInfo(
```
Comment (Mazel-Tovr) [inline core/src/main/kotlin/endpoints/AgentManager.kt]: InstanceState

File: core/src/main/kotlin/endpoints/AgentManager.kt [hunk 6]
```diff
@@ -549,10 +532,62 @@ class AgentManager(override val kodein: Kodein) : KodeinAware {
...
+            instances[instanceKey]?.let { instance ->
+                instance[instanceId]?.copy(status = status)?.let {
+                    logger.trace { "instance $instanceId changed status, new status is $status" }
+                    instances.put(instanceKey, instance.put(instanceId, it))
+                }
+            } ?: instances
+        }[instanceKey]
+    }
+
+    fun getActualAgentStatus(agentId: String): AgentStatus = instanceIds(agentId).let { instances ->
```
Comment (Mazel-Tovr) [inline core/src/main/kotlin/endpoints/AgentManager.kt]: getStatus

File: core/src/main/kotlin/endpoints/AgentManager.kt [hunk 7]
```diff
@@ -549,10 +532,62 @@ class AgentManager(override val kodein: Kodein) : KodeinAware {
...
-        notifySingleAgent(id)
+        notifySingleAgent(agentId)
+    }
+
+    private suspend fun wrapInstanceBusy(
+        agentId: String,
+        instanceId: String,
+        instanceInfo: InstanceInfo,
+        block: suspend AgentWsSession.() -> Unit
+    ) {
```
Comment (Mazel-Tovr) [inline core/src/main/kotlin/endpoints/AgentManager.kt]: reduce amount of param

**Review by LikhanovMaxim — APPROVED** [review]

File: core/src/main/kotlin/endpoints/AgentManager.kt [hunk 8]
```diff
@@ -549,10 +533,63 @@ class AgentManager(override val kodein: Kodein) : KodeinAware {
...
+            instances[agent]?.let { instance ->
+                logger.trace { "instance $instanceKey changed status, new status is $status" }
+                instances.put(agent, instance.put(instanceKey, instanceState.copy(status = status)))
+            } ?: instances
+        }[agent]
+    }
+
+    private fun getInstanceState(instanceKey: InstanceKey) = _instances.value[instanceKey.agentKey]?.get(instanceKey)
+
+    fun getStatus(agentId: String): AgentStatus = instanceIds(agentId).let { instances ->
```
Comment (ViktoriiaLebedeva) [inline core/src/main/kotlin/endpoints/AgentManager.kt:583]: Can you check this variant of these function 
`fun getStatus(agentId: String): AgentStatus = instanceIds(agentId).takeIf { it.isNotEmpty() }?.let { instances ->
        entryOrNull(agentId)?.agent?.takeIf { it.isRegistered }?.let {
            val statuses = instances.map { it.value.status }.distinct()
            AgentStatus.ONLINE.takeIf { AgentStatus.ONLINE in statuses } ?: AgentStatus.BUSY
        } ?: AgentStatus.NOT_REGISTERED
    } ?: AgentStatus.OFFLINE`
Offline status can be only if agent doesn't have instances.

File: core/src/main/kotlin/endpoints/AgentManager.kt [hunk 9]
```diff
@@ -549,10 +533,63 @@ class AgentManager(override val kodein: Kodein) : KodeinAware {
...
+        notifySingleAgent(agentId)
+    }
+
+    private suspend fun wrapInstanceBusy(
+        instanceKey: InstanceKey,
+        block: suspend AgentWsSession.() -> Unit,
+    ): Unit = getInstanceState(instanceKey)?.let { instanceState ->
+        val agentId = instanceKey.agentKey.agentId
+        updateInstanceStatus(instanceKey, AgentStatus.BUSY).also { instance ->
+            if (instance?.all { it.value.status == AgentStatus.BUSY } == true) {
```
Comment (ViktoriiaLebedeva) [inline core/src/main/kotlin/endpoints/AgentManager.kt:552]: simplify boolean expression to if(instance.all { it.value.status == AgentStatus.BUSY })

File: core/src/main/kotlin/endpoints/AgentManager.kt [hunk 10]
```diff
@@ -549,10 +533,63 @@ class AgentManager(override val kodein: Kodein) : KodeinAware {
...
-        notifySingleAgent(id)
+        notifySingleAgent(agentId)
+    }
+
+    private suspend fun wrapInstanceBusy(
+        instanceKey: InstanceKey,
+        block: suspend AgentWsSession.() -> Unit,
+    ): Unit = getInstanceState(instanceKey)?.let { instanceState ->
+        val agentId = instanceKey.agentKey.agentId
+        updateInstanceStatus(instanceKey, AgentStatus.BUSY).also { instance ->
```
Comment (ViktoriiaLebedeva) [inline core/src/main/kotlin/endpoints/AgentManager.kt:551]: Move the null check here, before also.

**Review by ViktoriiaLebedeva — CHANGES_REQUESTED** [review]

**Pull request merged**